x/programs/rust/README.md

```diff
@@ -1,56 +1,179 @@
+- Optionally use our build script.
 
-`./scripts/build.sh`
+```sh
+./scripts/build.sh
+```
 
-## Serialization Between Host(Go) & Guest(Rust)
+## Serialization Between Host and Program
 
-Just a couple things to note. Serialization is minimal, yet there are certain aspects in the code that need to follow a specific format. Specifically there are currently only two places we modify the bytes coming in/out of rust.
+Just a couple things to note. Serialization is minimal, yet there are certain
+aspects in the code that need to follow a specific format. Specifically, there
+are currently only two places we modify the bytes coming in/out of rust.
 
-- The first is quite minimal. When storing a value in the host, the `wasmlanche_sdk` prepends a single byte representing the type of `ProgramValue` being stored. This single byte is necessary to inform Rust about the variable type when retrieving from the `host`.
-- The second area is a bit more complex and happens during a call to invoke another program. In this case we pass a byte array which contains the parameters for the external function call. To construct this we marshal all the params with their metadata into one final byte array. Each parameter is added in this order
+1. The first is quite minimal. When storing a value in the host, the
+SDK prepends a single byte representing the type of `ProgramValue`
+being stored. This single byte is necessary to inform Rust about the variable
+type when retrieving from the `host`.
+2. The second area is a bit more complex and happens during a call to invoke
+another program. In this case, we pass a byte array which contains the parameters
+for the external function call. To construct this we marshal all the params with
+their metadata into one final byte array. Each parameter is added in this order
   - length of the parameter in bytes(stored as a i64)
   - boolean, [1] if the parameter is an Int, [0] otherwise
   - the actual bytes of the parameter.
 
 On the Go side, we unmarshal in the same order.
 
+## Storage
+
+A Program has 2 storage types.
+
+- `Memory` in WebAssembly is a linear buffer of unsigned bytes that can be read and
+written to by both the `Program` or host.
+
+- `State` is a host module which is imported into the `Program` instance. It
+exposes access to a database store. Unlike `Memory` state can be durable.
+
+### Memory
+
+In this example we share a byte slice between our host and a `Program`. To
+achieve this, we can write our byte slice to the shared memory from the host using
+an exported functions called `alloc` and `memory`. To do this, we would first
+allocate a chunk of continuous memory of the length desired.  `Alloc` will
+return the `offset` of that slice.
+
+```go
+runtime := // create new runtime and initialize
+memory := runtime.Memory()
+
+bytes := []byte("hello world")
+bytesLen := uint64(len(bytes))
+
+// create a memory allocation 
+offset, _ := memory.Alloc(bytesLen)
```

Comment:
Is there some error we should be checking here? (maybe out of scope for readme)

Reply:
yeah just example usage error removed for brevity.